eclipse-collections/src/main/java/org/eclipse/collections/impl/multimap/AbstractSynchronizedMultimap.java

```diff
@@ -0,0 +1,399 @@
+    {
+        synchronized (this.lock)
+        {
+            return this.getDelegate().putAll(key, values);
+        }
+    }
+
+    @Override
+    public <KK extends K, VV extends V> boolean putAll(Multimap<KK, VV> multimap)
+    {
+        synchronized (this.lock)
+        {
+            return this.getDelegate().putAll(multimap);
+        }
+    }
+
+    @Override
+    public void clear()
+    {
+        synchronized (this.lock)
+        {
+            this.getDelegate().clear();
+        }
+    }
+
+    @Override
+    public boolean isEmpty()
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.isEmpty();
+        }
+    }
+
+    @Override
+    public boolean notEmpty()
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.notEmpty();
+        }
+    }
+
+    @Override
+    public void forEachValue(Procedure<? super V> procedure)
+    {
+        synchronized (this.lock)
+        {
+            this.delegate.forEachValue(procedure);
+        }
+    }
+
+    @Override
+    public void forEachKey(Procedure<? super K> procedure)
+    {
+        synchronized (this.lock)
+        {
+            this.delegate.forEachKey(procedure);
+        }
+    }
+
+    @Override
+    public void forEachKeyValue(Procedure2<? super K, ? super V> procedure)
+    {
+        synchronized (this.lock)
+        {
+            this.delegate.forEachKeyValue(procedure);
+        }
+    }
+
+    @Override
+    public void forEachKeyMultiValues(Procedure2<? super K, ? super Iterable<V>> procedure)
+    {
+        synchronized (this.lock)
+        {
+            this.delegate.forEachKeyMultiValues(procedure);
+        }
+    }
+
+    @Override
+    public int size()
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.size();
+        }
+    }
+
+    @Override
+    public int sizeDistinct()
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.sizeDistinct();
+        }
+    }
+
+    @Override
+    public boolean containsKey(Object key)
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.containsKey(key);
+        }
+    }
+
+    @Override
+    public boolean containsValue(Object value)
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.containsValue(value);
+        }
+    }
+
+    @Override
+    public boolean containsKeyAndValue(Object key, Object value)
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.containsKeyAndValue(key, value);
+        }
+    }
+
+    @Override
+    public RichIterable<K> keysView()
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.keysView();
+        }
+    }
+
+    @Override
+    public SetIterable<K> keySet()
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.keySet();
+        }
```

Comment:
Could you add assertions for removing from the keySet(), in addition to
adding to it?
On Mon, Aug 15, 2016 at 9:22 AM Shotaro Sano <notifications@github.com>
wrote:

> In
> eclipse-collections/src/main/java/org/eclipse/collections/impl/multimap/AbstractSynchronizedMultimap.java
> <https://github.com/eclipse/eclipse-collections/pull/93#discussion_r74760107>
> :
>
> > +    @Override
> > +    public RichIterable<K> keysView()
> > +    {
> > +        synchronized (this.lock)
> > +        {
> > +            return this.delegate.keysView();
> > +        }
> > +    }
> > +
> > +    @Override
> > +    public SetIterable<K> keySet()
> > +    {
> > +        synchronized (this.lock)
> > +        {
> > +            return this.delegate.keySet();
> > +        }
>
> @motlin <https://github.com/motlin> @itohro <https://github.com/itohro>
>
> Thanks for your input!
>
> Casted #keySet result and wrap it in synchronized wrapper. The set also
> should be wrapped in unmodifiable one implicitly because of Multimap#keymap
> behavior.
>
> @Overridepublic SetIterable<K> keySet()
> {
>     synchronized (this.lock)
>     {
>         return SynchronizedMutableSet.of((Set<K>) this.getDelegate().keySet(), this.lock);
>     }
> }
>
> Replaced a line in AbstractMultimapTestCase#keySet so that it checks
> UnsupportedOperationException for mutation, instead of asserting keySet’s
> class.
>
> @Testpublic void keySet()
> {
>     Multimap<String, Integer> multimap =
>             this.newMultimapWithKeysValues("One", 1, "One", 1, "Two", 2, "Three", 3);
>     Set<String> keySet = (Set<String>) multimap.keySet();
>     Verify.assertThrows(UnsupportedOperationException.class, () -> keySet.add("Four"));
>     Assert.assertEquals(Sets.mutable.of("One", "Two", "Three"), keySet);
> }
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse/eclipse-collections/pull/93/files/ace474519c1dbe03d2783ec5640c911f58ada19d..7de3f8ba027df68f6a4eb3621877b9007dbe0c11#r74760107>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAO6IixuL0bTsIbTH7XGyQHp6OxQ-tt1ks5qgGgxgaJpZM4ILYAA>
> .
>